### PR TITLE
Filter out Abstract Actors + Use Constant for Support Directory

### PR DESCRIPTION
**Pull request by PeterAntal**
This exercises @pchote suggestion of skipping partial templates, to avoid need for compensating for errors on incomplete cases, which is inelegant.

This approach ignores all abstract actor templates, in MergeOrDefault(...) process.

As part of the change I added a constant for the support directory symbol, to provide semantic context for "^".

Testing revealed no regressions - If there's particular arrangements of data on rulesets which could be adversely affected, I'd be happy to sanity check.

### CODE REVIEW

**Comment by GraionDilach** [comment]
You broke the unit tests.

**Comment by GraionDilach** [comment]
Let me try to understand the point of this PR - if this PR is merged, then templates themselves don't end up treated as additional ActorInfo entries in the ruleset?

File: OpenRA.Game/GameRules/Ruleset.cs [hunk 1]
```diff
@@ -87,16 +87,19 @@ public class Ruleset
 
 		public IEnumerable<KeyValuePair<string, MusicInfo>> InstalledMusic { get { return Music.Where(m => m.Value.Exists); } }
 
-		static IReadOnlyDictionary<string, T> MergeOrDefault<T>(string name, IReadOnlyFileSystem fileSystem, IEnumerable<string> files, MiniYaml additional,
-			IReadOnlyDictionary<string, T> defaults, Func<MiniYamlNode, T> makeObject)
+		static IReadOnlyDictionary<string, T> MergeOrDefault<T>(string name,
```
Comment (pchote) [inline OpenRA.Game/GameRules/Ruleset.cs]: IMO a cleaner approach here would be to pass in a `Func<MiniYamlNode, bool> filterNode` argument, then have the callers pass in either `n => .StartsWith("^", StringComparison.Ordinal)` (for actors) or `_ => true` (everything else).

File: OpenRA.Game/GameRules/Ruleset.cs [hunk 2]
```diff
@@ -261,5 +267,14 @@ static bool AnyFlaggedTraits(ModData modData, List<MiniYamlNode> actors)
...
 			return false;
 		}
+
+		/// <summary>
+		/// Filter for concrete Nodes.
+		/// Abstract nodes (e.g. rooted from support directory) are not accepted.
+		/// </summary>
+		private static bool IsConcreteNode(MiniYamlNode node)
+		{
+			return !SymbolicSupportDirectory.IsPathRelativeToSupportDirectory(node.Key);
```
Comment (pchote) [inline OpenRA.Game/GameRules/Ruleset.cs]: The `^` convention for actor names is completely separate to the convention for using `^` to denote the support directory in file paths.  Mixing the two here isn't a good idea!

Comment (PeterAntal) [reply OpenRA.Game/GameRules/Ruleset.cs]: @pchote - Yeah, in hindsight I should have just gone with a minimal change on the impl. 

How would you feel about definition of "abstractActorPrefix" string constant on ActorInfo to disambiguate that case, vs dropping the constants business and sticking with the string literals?

**Comment by PeterAntal** [comment]
@GraionDilach - Yes, to your question - This change is about avoiding exception handling or later pre-emption of errors during handling of the template actors.

File: OpenRA.Game/GameRules/Ruleset.cs (hunk 1 above)
Comment (PeterAntal) [reply OpenRA.Game/GameRules/Ruleset.cs]: Makes sense. I went with a default behavior of no-op rather than requiring explicit behavior on the other callers.

File: OpenRA.Game/GameRules/Ruleset.cs [hunk 3]
```diff
@@ -108,7 +117,8 @@ public static Ruleset LoadDefaults(ModData modData)
 			Action f = () =>
 			{
 				var actors = MergeOrDefault("Manifest,Rules", fs, m.Rules, null, null,
-					k => new ActorInfo(modData.ObjectCreator, k.Key.ToLowerInvariant(), k.Value));
+					k => new ActorInfo(modData.ObjectCreator, k.Key.ToLowerInvariant(), k.Value),
+					filterNode: n => n.Key.StartsWith(ActorInfo.AbstractActorPrefix, StringComparison.Ordinal));
```
Comment (PeterAntal) [inline OpenRA.Game/GameRules/Ruleset.cs]: I went with explicit param identifier for readability, as the actor case is the only one opting into the filter pattern, and I didn't want to spam the others with boilerplate.

File: OpenRA.Game/GameRules/Ruleset.cs [hunk 4]
```diff
@@ -87,16 +87,26 @@ public class Ruleset
...
 				return defaults;
 
-			var result = MiniYaml.Load(fileSystem, files, additional)
-				.ToDictionaryWithConflictLog(k => k.Key.ToLowerInvariant(), makeObject, "LoadFromManifest<" + name + ">");
+			var yamlNodes = MiniYaml.Load(fileSystem, files, additional) as IEnumerable<MiniYamlNode>;
 
-			return new ReadOnlyDictionary<string, T>(result);
+			// Optionally, the caller can filter out elements from the loaded set of nodes. Default behavior is unfiltered.
+			if (filterNode != null)
+			{
```
Comment (abcdefg30) [inline OpenRA.Game/GameRules/Ruleset.cs]: No braces around single-line statements, please. ^^

File: OpenRA.Mods.Common/UtilityCommands/CheckSequenceSprites.cs [hunk 5]
```diff
@@ -39,7 +39,7 @@ void IUtilityCommand.Run(Utility utility, string[] args)
 				Console.WriteLine("Tileset: " + ts.Name);
 				var sc = new SpriteCache(modData.DefaultFileSystem, modData.SpriteLoaders, new SheetBuilder(SheetType.Indexed));
 				var nodes = MiniYaml.Merge(modData.Manifest.Sequences.Select(s => MiniYaml.FromStream(modData.DefaultFileSystem.Open(s), s)));
-				foreach (var n in nodes.Where(node => !node.Key.StartsWith("^")))
+				foreach (var n in nodes.Where(node => !node.Key.StartsWith(SymbolicSupportDirectory.Identifier)))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/UtilityCommands/CheckSequenceSprites.cs]: This looks like we're mixing directory and actor name identifiers. (?)

Comment (PeterAntal) [reply OpenRA.Mods.Common/UtilityCommands/CheckSequenceSprites.cs]: That's right- actors are on those nodes. Fixed.

File: OpenRA.Game/GameRules/Ruleset.cs (hunk 4 above)
Comment (PeterAntal) [reply OpenRA.Game/GameRules/Ruleset.cs]: Thanks - old habit. I'll see if I can find a linter or editor extension to help me catch these.

File: OpenRA.Game/SymbolicSupportDirectory.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+#endregion
+
+using System;
+
+namespace OpenRA
+{
+	/// <summary>
+	/// Responsible for constants and common logic of dealing with the Symbolic support directory path.
+	/// </summary>
+	public class SymbolicSupportDirectory
```
Comment (pchote) [inline OpenRA.Game/SymbolicSupportDirectory.cs]: Please split these support directory changes into their own commit.
OpenRA conventions are to keep distinct changes in distinct commits!

File: OpenRA.Game/SymbolicSupportDirectory.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	{
+		public const string Identifier = "^";
```
Comment (pchote) [inline OpenRA.Game/SymbolicSupportDirectory.cs]: This would be better defined as `public const string SupportDirPrefix` on the `Platform` class.

File: OpenRA.Game/SymbolicSupportDirectory.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+		public static bool IsPathRelativeToSupportDirectory(string path)
```
Comment (pchote) [inline OpenRA.Game/SymbolicSupportDirectory.cs]: This would also be better in `Platform`, which is where the rest of our support dir path related code lives.

**Review by pchote — COMMENTED** [review]
Looks sensible overall, but I haven't tested it.  Just a couple of minor things with the support dir changes:

File: OpenRA.Game/SymbolicSupportDirectory.cs (hunk 6 above)
Comment (PeterAntal) [reply OpenRA.Game/SymbolicSupportDirectory.cs]: Ack'ed-  I'll separate this facet to a separate PR.

File: OpenRA.Game/SymbolicSupportDirectory.cs (hunk 7 above)
Comment (PeterAntal) [reply OpenRA.Game/SymbolicSupportDirectory.cs]: Ack'ed - Makes sense.

**Comment by PeterAntal** [comment]
PR is updated to omit support Dir Prefix aspects, which have been moved to separate PR.

**Review by pchote — DISMISSED** [review]
LGTM :+1:

Not strictly a blocker, but it would be good if you could change the first line of the commit message to just "Allow each caller on MergeOrDefault to discard nodes" so that it is not truncated in the git shortlog.

**Comment by PeterAntal** [comment]
Updated the commit message per suggestion.

File: OpenRA.Mods.Common/UtilityCommands/CheckSequenceSprites.cs [hunk 9]
```diff
@@ -39,7 +39,7 @@ void IUtilityCommand.Run(Utility utility, string[] args)
 				Console.WriteLine("Tileset: " + ts.Name);
 				var sc = new SpriteCache(modData.DefaultFileSystem, modData.SpriteLoaders, new SheetBuilder(SheetType.Indexed));
 				var nodes = MiniYaml.Merge(modData.Manifest.Sequences.Select(s => MiniYaml.FromStream(modData.DefaultFileSystem.Open(s), s)));
-				foreach (var n in nodes.Where(node => !node.Key.StartsWith("^")))
+				foreach (var n in nodes.Where(node => !node.Key.StartsWith(ActorInfo.AbstractActorPrefix)))
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/UtilityCommands/CheckSequenceSprites.cs]: Specify `StringComparison.Ordinal` here.

File: OpenRA.Game/GameRules/Ruleset.cs [hunk 10]
```diff
@@ -87,16 +87,24 @@ public class Ruleset
...
+			IReadOnlyDictionary<string, T> defaults,
+			Func<MiniYamlNode, T> makeObject,
+			Func<MiniYamlNode, bool> filterNode = null)
 		{
 			if (additional == null && defaults != null)
 				return defaults;
 
-			var result = MiniYaml.Load(fileSystem, files, additional)
-				.ToDictionaryWithConflictLog(k => k.Key.ToLowerInvariant(), makeObject, "LoadFromManifest<" + name + ">");
+			var yamlNodes = MiniYaml.Load(fileSystem, files, additional) as IEnumerable<MiniYamlNode>;
```
Comment (RoosterDragon) [inline OpenRA.Game/GameRules/Ruleset.cs]: `IEnumerable<MiniYamlNode> yamlNodes = MiniYaml.Load(fileSystem, files, additional);` would suffice.

**Comment by abcdefg30** [comment]
Can you squash the commits, please?

**Comment by PeterAntal** [comment]
Commits squashed :)

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ddc7f8dff7d3a00faa0d500db04e82428aa5333d...3bf03e6acae3d6f9720011cab1e8705bb00e351b)
